revoke/revoke.go

```diff
@@ -123,69 +213,141 @@ func getIssuer(cert *x509.Certificate) *x509.Certificate {
 	}
 
 	return issuer
-
 }
 
-// check a cert against a specific CRL. Returns the same bool pair
-// as revCheck.
-func certIsRevokedCRL(cert *x509.Certificate, url string) (revoked, ok bool) {
-	crl, ok := CRLSet[url]
+// checks whether CRL in memory is valid
+func (r *Revoke) isInMemoryCRLValid(key string) bool {
+	defer r.lock.Unlock()
+	r.lock.Lock()
+	crl, ok := r.crlSet[key]
 	if ok && crl == nil {
 		ok = false
-		delete(CRLSet, url)
+		delete(r.crlSet, key)
 	}
 
-	var shouldFetchCRL = true
-	if ok {
-		if !crl.HasExpired(time.Now()) {
-			shouldFetchCRL = false
+	if ok && !crl.HasExpired(time.Now()) {
+		return true
+	}
+
+	return false
+}
+
+// fetchLocalCRL reads CRL from the local filesystem
+// force flag allows you to update the CRL
+func (r *Revoke) fetchLocalCRL(newLocalCRL string, force bool) error {
+	if force || !r.isInMemoryCRLValid(newLocalCRL) {
+		if _, err := os.Stat(newLocalCRL); err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+
+		tmp, err := ioutil.ReadFile(newLocalCRL)
+		if err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+		crl, err := x509.ParseCRL(tmp)
+		if err != nil {
+			return fmt.Errorf("failed to parse local CRL file: %v", err)
 		}
+
+		r.lock.Lock()
+		r.crlSet[newLocalCRL] = crl
+		if r.localCRL != newLocalCRL {
+			if _, ok := r.crlSet[r.localCRL]; ok {
+				delete(r.crlSet, r.localCRL)
+			}
+			r.localCRL = newLocalCRL
+		}
+		r.lock.Unlock()
 	}
 
+	return nil
+}
+
+// FetchRemoteCRL fetches remote CRL into internal map,
+// force overwrites previously read CRL
+func (r *Revoke) FetchRemoteCRL(url string, cert *x509.Certificate, force bool) error {
+	shouldFetchCRL := !r.isInMemoryCRLValid(url)
+
 	issuer := getIssuer(cert)
 
-	if shouldFetchCRL {
+	if force || shouldFetchCRL {
 		var err error
-		crl, err = fetchCRL(url)
+		crl, err := fetchCRL(url)
 		if err != nil {
-			log.Warningf("failed to fetch CRL: %v", err)
-			return false, false
+			return fmt.Errorf("failed to fetch CRL: %v", err)
 		}
 
 		// check CRL signature
 		if issuer != nil {
 			err = issuer.CheckCRLSignature(crl)
 			if err != nil {
-				log.Warningf("failed to verify CRL: %v", err)
-				return false, false
+				return fmt.Errorf("failed to verify CRL: %v", err)
 			}
 		}
 
-		CRLSet[url] = crl
+		r.lock.Lock()
+		r.crlSet[url] = crl
+		r.lock.Unlock()
+	}
+
+	return nil
+}
+
+// check a cert against a specific CRL. Returns the same bool pair
+// as revCheck. If remote is false - will assume that url is a file.
+func (r *Revoke) certIsRevokedCRL(cert *x509.Certificate, url string, remote bool) (revoked, ok bool) {
+	var err error
+	if remote {
+		err = r.FetchRemoteCRL(url, cert, false)
```

Comment:
UPD: Not fixed. Because we have to check path before update the value. And if I'd return CRL here, I have to implement additional checks in other functions, which will make the code more complicated (take a look at `SetLocalCRL`).